OpenBLAS support broken due to extern C linkage

##### System information (version)
- OpenCV => 3.4.0
- Operating System / Platform => Ubuntu 16.04
- Compiler => GCC 5.4

##### Detailed description
Hi!

we build OpenCV with OpenBLAS 0.2.20 (also our own build) ...ah, up to patch #9955, LAPACK detection had been working properly.

Starting with the changes in patch #9955 in the 3.3.1 release (and 3.4.0 too, to which we're trying to upgrade to atm), the LAPACK cmake detection script outputs 

> -- LAPACK(OpenBLAS): LAPACK_LIBRARIES: /usr/local/lib/libopenblas.so
> -- LAPACK(OpenBLAS): Can't build LAPACK check code. This LAPACK version is not supported.

I was able to track it down to the extern "C" encapsulation of the OpenBLAS headers in the lapack proxy which seems similar to issue #10230. Now, if we're removing the `extern "C"` statement, everything works fine again. 

I'm just wondering if it this is due to some trouble with our own build version of OpenBLAS as #9955 was related to the OpenBLAS support in OpenCV, too. @alalek Why did you put the OpenBLAS headers in extern C linkage?

##### Workaround

Atm, I'd removed the extern C linkage from the detection script to build OpenCV with OpenBLAS...


Thanks in advance and regards

Ah, I'm aware of the LAPACK/BLAS origins and the correct usage of extern "C" :) - sorry if my post was irritating...but thanks for the SA link, the std reference for nested linkage is nice :+1: 
I already tried nm before posting...all symbols are (as expected) easily readable C-symbols, no C++ name mangling there, sorry, forgot adding it. 

I don't know what apple is doing in their framework, I was just guessing from #10230 comments and the differences in the CMake detection script.

Ok, I tracked down the error a bit more...fortunately, there was an error output in the `OpenCVFindLAPACK.cmake` :)

It seems, the error comes from the usage of GCC complex number library:

>  In file included from /usr/include/c++/5/complex:42:0,
>
>                 from /usr/include/c++/5/ccomplex:38,
>                 from /usr/include/c++/5/complex.h:32,
>                 from /usr/local/include/openblas/lapacke.h:73,
>                 from /tmp/opencv-3.4.0/release/opencv_lapack.h:3,
>                 from /tmp/opencv-3.4.0/cmake/checks/lapack_check.cpp:1:
> 
>   /usr/include/c++/5/bits/cpp_type_traits.h:72:3: error: template with C linkage
>
>    template<typename _Iterator, typename _Container>
> ...
> ...

I've set `ENABLE_CXX11=ON` in my configure options, which forces try_compile to set the compile std to C++11, thus complex.h includes ccomplex which requires templates which are non-existent in C :)

> \#if __cplusplus >= 201103L
> \#include \<ccomplex\>
> \#endif

So, forcing try_compile to use C++98 for the test compilation fixes the configure bug. But that's a bit of an ugly hack. 

So, sorry. I'm still not getting it - why are you putting a library which already has C-linkage for it's symbols into another nested C-linkage layer? And thereby putting the std library headers into extern linkage blocks, from which they're intentionally excluded! See https://github.com/xianyi/OpenBLAS/pull/490

I could provide a patch for the OpenCVFindLAPACK.cmake if you want...

Thank you!

--- a/cmake/OpenCVFindLAPACK.cmake
+++ b/cmake/OpenCVFindLAPACK.cmake
@@ -31,7 +31,7 @@ macro(ocv_lapack_check)
   else()
     # adding proxy opencv_lapack.h header
     set(CBLAS_H_PROXY_PATH ${CMAKE_BINARY_DIR}/opencv_lapack.h)
-    if(APPLE)
+    if((APPLE OR OPENCV_SKIP_LAPACK_EXTERN_C) AND NOT OPENCV_FORCE_LAPACK_EXTERN_C)
         set(_lapack_include_str_extern_C "")
         set(_lapack_include_str_extern_C_end "")
     else()

--- a/cmake/checks/lapack_check.cpp
+++ b/cmake/checks/lapack_check.cpp
@@ -1,3 +1,4 @@
+#include <complex.h>
 #include "opencv_lapack.h"
 
 static char* check_fn1 = (char*)sgesv_;

--- a/modules/core/src/hal_internal.cpp
+++ b/modules/core/src/hal_internal.cpp
@@ -46,6 +46,7 @@
 
 #ifdef HAVE_LAPACK
 
+#include <complex.h>
 #include "opencv_lapack.h"
 
 #include <cmath>

